sw/device/silicon_creator/lib/drivers/alert.c

```diff
@@ -0,0 +1,175 @@
+                       2);
+      abs_mmio_write32(
+          kBase + ALERT_HANDLER_LOC_ALERT_CLASS_0_REG_OFFSET + index, 2);
+      break;
+    case kAlertClassD:
+      abs_mmio_write32(kBase + ALERT_HANDLER_ALERT_CLASS_0_REG_OFFSET + index,
+                       3);
+      abs_mmio_write32(
+          kBase + ALERT_HANDLER_LOC_ALERT_CLASS_0_REG_OFFSET + index, 3);
+      break;
+    case kAlertClassX:
+      return kErrorOk;
+    default:
+      return kErrorAlertBadClass;
+  }
+
+  switch (enabled) {
+    case kAlertEnableNone:
+      break;
+    case kAlertEnableLocked:
+      // Enable, then lock.
+      abs_mmio_write32(kBase + ALERT_HANDLER_ALERT_EN_0_REG_OFFSET + index, 1);
+      abs_mmio_write32(kBase + ALERT_HANDLER_LOC_ALERT_EN_0_REG_OFFSET + index,
+                       1);
+      abs_mmio_write32(kBase + ALERT_HANDLER_ALERT_REGWEN_0_REG_OFFSET + index,
+                       0);
+      abs_mmio_write32(
+          kBase + ALERT_HANDLER_LOC_ALERT_REGWEN_0_REG_OFFSET + index, 0);
+      break;
+    case kAlertEnableEnabled:
+      abs_mmio_write32(kBase + ALERT_HANDLER_ALERT_EN_0_REG_OFFSET + index, 1);
+      abs_mmio_write32(kBase + ALERT_HANDLER_LOC_ALERT_EN_0_REG_OFFSET + index,
+                       1);
+      break;
+    default:
+      return kErrorAlertBadEnable;
+  }
+
+  return kErrorOk;
+}
+
+rom_error_t alert_class_configure(alert_class_t cls,
+                                  const alert_class_config_t *config) {
+  uint32_t offset = 0;
+  uint32_t reg = 0;
+
+  // Each escalation signall should be asserted in its corresponding phase.
+  reg = bitfield_field32_write(reg, ALERT_HANDLER_CLASSA_CTRL_MAP_E0_FIELD, 0);
+  reg = bitfield_field32_write(reg, ALERT_HANDLER_CLASSA_CTRL_MAP_E1_FIELD, 1);
+  reg = bitfield_field32_write(reg, ALERT_HANDLER_CLASSA_CTRL_MAP_E2_FIELD, 2);
+  reg = bitfield_field32_write(reg, ALERT_HANDLER_CLASSA_CTRL_MAP_E3_FIELD, 3);
+
+  // All of the alert class register blocks are identical but at different
+  // offsets.  We'll treat everything like Class A, but add in the offset
+  // to the other classes.
+  switch (cls) {
+    case kAlertClassA:
+      offset = ALERT_HANDLER_CLASSA_CTRL_REG_OFFSET -
+               ALERT_HANDLER_CLASSA_CTRL_REG_OFFSET;
+      break;
+    case kAlertClassB:
+      offset = ALERT_HANDLER_CLASSB_CTRL_REG_OFFSET -
+               ALERT_HANDLER_CLASSA_CTRL_REG_OFFSET;
+      break;
+    case kAlertClassC:
+      offset = ALERT_HANDLER_CLASSC_CTRL_REG_OFFSET -
+               ALERT_HANDLER_CLASSA_CTRL_REG_OFFSET;
+      break;
+    case kAlertClassD:
+      offset = ALERT_HANDLER_CLASSD_CTRL_REG_OFFSET -
+               ALERT_HANDLER_CLASSA_CTRL_REG_OFFSET;
+      break;
+    case kAlertClassX:
+    default:
+      return kErrorAlertBadClass;
+  }
+  switch (config->enabled) {
+    case kAlertEnableNone:
+      break;
+    case kAlertEnableLocked:
+      reg = bitfield_bit32_write(reg, ALERT_HANDLER_CLASSA_CTRL_LOCK_BIT, true);
```

Comment:
Do we want to lock the escalation counter only for classes whose configurations are locked or for all alerts? Asking because I was expecting `kAlertEnableLocked` to affect only `WEN` registers. We could also expose this as a separate field in `alert_class_config_t` but I'm not sure if we have a use-case.

Reply:
I agree, apologies for not being clear: I was thinking maybe we should lock the escalation counter for both `kAlertEnableLocked` and `kAlertEnableEnabled` but I guess that's not what we want since `kAlertEnableEnabled` is not supposed to lock.